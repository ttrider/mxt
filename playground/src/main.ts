import { observable, computed } from "mobx"

import { ex01 } from "./templates/ex01";
import { ex02 } from "./templates/ex02";
import { ex03 } from "./templates/ex03";

class TestData {
    @observable r = 200;
    @observable g = 18;
    @observable b = 228;

    @computed get color() {
        return "#" + this.r.toString(16) + this.g.toString(16) + this.b.toString(16);
    }

    colorClick() {

        this.r += 20;
        this.g += 15;
        this.b += 35;

        this.r = this.r % 255;
        this.g = this.g % 255;
        this.b = this.b % 255;

        if (this.el){
            //this.el.dispose();

            const root_mv = document.getElementById("mv");
            if (root_mv) {
                this.el.appendTo(root_mv);
            }
        }
    }

    el:any
}


const data = new TestData();

if (document) {

    // example 01
    const root_ex01 = document.getElementById("ex01");
    if (root_ex01) {

        data.el = ex01(data);
        data.el.appendTo(root_ex01);

    }

    // example 02
    const root_ex02 = document.getElementById("ex02");
    if (root_ex02) {
        ex02(data).appendTo(root_ex02);
    }

    // example 03
    const root_ex03 = document.getElementById("ex03");
    ex03(data, root_ex03);
}

