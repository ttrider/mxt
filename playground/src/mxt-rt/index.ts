export * from "./types";
export * from "./context";
export * from "./template-context";
export * from "./data-context";
export * from "./parts-context";
export * from "./loop-context";
export * from "./styles-context";