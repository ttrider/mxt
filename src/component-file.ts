import { ElementInfo, StyleElementInfo, TemplateInfo } from ".";
import { parseDOM } from "htmlparser2";
import * as dom from "domhandler";
import fs from "fs";
import path from "path";
import util from "util";
import ts from "typescript";
import * as d from "./ts";

const readFile = util.promisify(fs.readFile);


export class ComponentFile {

    filePath: string;
    content: string;
    dom: dom.Node[];
    links: ElementInfo[] = [];
    globalStyles: StyleElementInfo[] = [];

    templates: { [id: string]: TemplateInfo } = {};
    errors: Error[] = [];

    initStatements = d.StatementList();
    componentStatements = d.StatementList();

    importStatements: {
        [from: string]: { name: string, as: string }[]
    } = {};

    private constructor(content: string, filePath: string) {

        this.content = content;
        this.dom = parseDOM(this.content, { xmlMode: true, withStartIndices: true, withEndIndices: true });
        this.filePath = filePath;
    }

    addImport(identifier: string | { name: string, as: string } | Array<string | { name: string, as: string }>, from: string) {

        if (!Array.isArray(identifier)) {
            identifier = [identifier];
        }
        const definitions = identifier.map<{ name: string, as: string }>(i => {
            if (typeof i === "string") {
                i = { name: i, as: i };
            }
            return i;
        });


        const fromSet = this.importStatements[from];
        if (!fromSet) {
            this.importStatements[from] = definitions;
            return;
        }

        for (const definition of definitions) {
            // lookup existing definition
            const existing = fromSet.find((item) => item.as === definition.as);
            if (!existing) {
                fromSet.push(definition);
            } else {

                if (definition.name != existing.name) {
                    throw new Error(`redefinition of the import: '${existing.name}' by ${definition.name}`);
                }
            }
        }
        return this;
    }

    getImportStatements() {

        const items: ts.Statement[] = [];

        for (const from in this.importStatements) {
            if (this.importStatements.hasOwnProperty(from)) {

                const ii = d.ImportStatement(this.importStatements[from], from);

                items.push(ii);
            }
        }
        return items;
    }

    getStatements() {
        const s =
            [
                ...this.getImportStatements(),
                ...this.initStatements,
                ...this.componentStatements
            ] as ts.Statement[];
        return s;
    }






    static fromContent(content: string) {
        return new ComponentFile(content, "./dummy.html")
    }

    static async fromFile(filePath: string) {
        filePath = path.resolve(filePath);
        const fileBuffer = await readFile(filePath);
        if (!fileBuffer) {
            throw new Error("can't read the file: " + filePath);
        }
        const content = fileBuffer.toString();
        return new ComponentFile(content, filePath);
    }
}
