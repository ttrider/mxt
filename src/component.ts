import { TokenizedContent, TemplateInfo } from ".";
import { Part, PartReference } from "./template-parts";

export class Component {

    static partIndex: number = 0;
    static newPartId() {
        return `p${Component.partIndex++}`;
    }


    dynamicStyles?: TokenizedContent[];
    styles?: TokenizedContent[];
    id: string;
    rootPart?: PartReference;
    parts: { [id: string]: TemplateInfo } = {};

    importParts: { [id: string]: string } = {};

    partset: { [id: string]: Part } = {};

    constructor(id: string) {
        this.id = id;
    }

    addComponentImport(name: string) {
        if (this.importParts[name]) {
            return this.importParts[name];
        }
        return this.importParts[name] = Component.newPartId();
    }

    newPart<T>(init?: T, id?: string): Part & T {
        if (!id) {
            id = Component.newPartId();
        }

        const part = (init) ? {
            id,
            ...init
        } as Part & T : { id } as Part & T;

        this.partset[id] = part;

        return part;
    }

}