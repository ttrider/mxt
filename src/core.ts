import { ElementInfo, StyleElementInfo, TemplateInfo } from ".";
import { Statement } from "@babel/types";
import * as dom from "domhandler";


export interface ComponentFile {

    path: string,
    name: string,
    content: string,
    dom: dom.Node[],
    links: ElementInfo[],
    globalStyles: StyleElementInfo[],

    templates: { [id: string]: TemplateInfo },
    errors: Error[],

    initStatements: Statement[],
    componentStatements: Statement[],
}
