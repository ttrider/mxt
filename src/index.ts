import glob from "glob";
import util from "util";
import fs from "fs";
import path from "path";
import { parseDOM } from "htmlparser2";
import { Node, Element } from "domhandler";
import { ElementType } from "domelementtype";
import { Handler } from "htmlparser2/lib/Parser";

const globp = util.promisify(glob);
const readFile = util.promisify(fs.readFile);

import { parseSync, parse, traverse } from "@babel/core";
import { statement } from "@babel/template";
import { Expression, SpreadElement, Statement, Scopable, V8IntrinsicIdentifier, JSXNamespacedName, ArgumentPlaceholder, VariableDeclaration } from "@babel/types";





declare type PluginHandlers =
    "beforeLoad" |
    "afterLoad" |
    "beforePackage" |
    "package" |
    "afterPackage" |
    "done";

declare type PluginComponentHandlers =
    "beforeCodegen" |
    "codegen" |
    "afterCodegen";


interface PluginHandlerMap {
    "beforeTransform": Node;
    "transform": Node;
    "afterTransform": Node;

    "beforeTransformStyle": Element;
    "transformStyle": Element;
    "afterTransformStyle": Element;

    "beforeTransformLink": Element;
    "transformLink": Element;
    "afterTransformLink": Element;

    "beforeTransformScript": Element;
    "transformScript": Element;
    "afterTransformScript": Element;

    "beforeTransformTemplate": Element;
    "transformTemplate": Element;
    "afterTransformTemplate": Element;

    "beforeParseStyle": Element;
    "parseStyle": Element;
    "afterParseStyle": Element;

    "beforeParseScript": Element;
    "parseScript": Element;
    "afterParseScript": Element;

    "beforeParseTemplate": Element;
    "parseTemplate": Element;
    "afterParseTemplate": Element;
}

export interface ElementInfo {
    name: string,
    attributes: { [name: string]: string },
    content?: string,
    expressions?: string[];

}
export interface StyleElementInfo extends ElementInfo {
    rules: string[]
}

export interface TemplateInfo extends ElementInfo {
    id: string,
    elements: Element[],
    tokens: AttributeTokenInfo[];
}

export interface ExpressionInfo {
    content: string,
    hasTokens: boolean,
    externalReferences: string[]
}

export interface AttributeTokenInfo extends ExpressionInfo {
    elementId: string;
    attributeName: string;
}

export interface Options {
    input: string;

    plugins?: Plugin[];

    //es5 vs es6

}

interface Plugin {
    beforeLoad?: (context: HandlerContext) => boolean;
    afterLoad?: (context: HandlerContext) => boolean;


    beforeTransform?: (context: HandlerContext, componentFile: ComponentFileInfo, node: Node) => boolean;
    transform?: (context: HandlerContext, componentFile: ComponentFileInfo, node: Node) => boolean;
    afterTransform?: (context: HandlerContext, componentFile: ComponentFileInfo, node: Node) => boolean;

    beforeTransformStyle?: (context: HandlerContext, componentFile: ComponentFileInfo, element: Element) => boolean;
    transformStyle?: (context: HandlerContext, componentFile: ComponentFileInfo, element: Element) => boolean;
    afterTransformStyle?: (context: HandlerContext, componentFile: ComponentFileInfo, element: Element) => boolean;

    beforeTransformLink?: (context: HandlerContext, componentFile: ComponentFileInfo, element: Element) => boolean;
    transformLink?: (context: HandlerContext, componentFile: ComponentFileInfo, element: Element) => boolean;
    afterTransformLink?: (context: HandlerContext, componentFile: ComponentFileInfo, element: Element) => boolean;

    beforeTransformScript?: (context: HandlerContext, componentFile: ComponentFileInfo, element: Element) => boolean;
    transformScript?: (context: HandlerContext, componentFile: ComponentFileInfo, element: Element) => boolean;
    afterTransformScript?: (context: HandlerContext, componentFile: ComponentFileInfo, element: Element) => boolean;

    beforeTransformTemplate?: (context: HandlerContext, componentFile: ComponentFileInfo, element: Element) => boolean;
    transformTemplate?: (context: HandlerContext, componentFile: ComponentFileInfo, element: Element) => boolean;
    afterTransformTemplate?: (context: HandlerContext, componentFile: ComponentFileInfo, element: Element) => boolean;

    beforeParseStyle?: (context: HandlerContext, componentFile: ComponentFileInfo, element: Element) => boolean;
    parseStyle?: (context: HandlerContext, componentFile: ComponentFileInfo, element: Element) => boolean;
    afterParseStyle?: (context: HandlerContext, componentFile: ComponentFileInfo, element: Element) => boolean;

    beforeParseScript?: (context: HandlerContext, componentFile: ComponentFileInfo, element: Element) => boolean;
    parseScript?: (context: HandlerContext, componentFile: ComponentFileInfo, element: Element) => boolean;
    afterParseScript?: (context: HandlerContext, componentFile: ComponentFileInfo, element: Element) => boolean;

    beforeParseTemplate?: (context: HandlerContext, componentFile: ComponentFileInfo, element: Element) => boolean;
    parseTemplate?: (context: HandlerContext, componentFile: ComponentFileInfo, element: Element) => boolean;
    afterParseTemplate?: (context: HandlerContext, componentFile: ComponentFileInfo, element: Element) => boolean;

    beforeCodegen?: (context: HandlerContext, componentFile: ComponentFileInfo) => boolean;
    codegen?: (context: HandlerContext, componentFile: ComponentFileInfo) => boolean;
    afterCodegen?: (context: HandlerContext, componentFile: ComponentFileInfo) => boolean;

    beforePackage?: (context: HandlerContext) => boolean;
    package?: (context: HandlerContext) => boolean;
    afterPackage?: (context: HandlerContext) => boolean;
    done?: (context: HandlerContext) => boolean;
}

export interface HandlerContext {

    options: Options,
    files: string[],
    componentFiles: ComponentFileInfo[],
    plugins: Plugin[],

    globalLinkElements: ElementInfo[],
    globalStyleElements: ElementInfo[],
    bodyElements: ElementInfo[]
}


export interface ComponentFileInfo {

    path: string,
    name: string,
    content: string,
    dom: Node[],
    links: ElementInfo[],
    globalStyles: StyleElementInfo[],

    templates: { [id: string]: TemplateInfo },
    errors: Error[]
}



async function mxt(options: Options) {

    options.plugins = (options.plugins || []);

    const files = await globp(options.input, { absolute: true });

    const context: HandlerContext = {
        options: options,
        files,
        plugins: (options.plugins || []),
        componentFiles: [],
        globalLinkElements: [],
        globalStyleElements: [],
        bodyElements: []
    };

    execute("beforeLoad", context);

    await loadFiles(context);

    execute("afterLoad", context);

    // transform stage
    for (const componentFile of context.componentFiles) {
        transformFile(context, componentFile);
    }
    // parse stage
    for (const componentFile of context.componentFiles) {
        parseFile(context, componentFile);
    }
    // codegen stage
    for (const componentFile of context.componentFiles) {
        codegenFile(context, componentFile);
    }

    execute("beforePackage", context);
    execute("package", context);
    execute("afterPackage", context);
    execute("done", context);
}
export default mxt;

async function loadFiles(context: HandlerContext) {

    for (const filePath of context.files) {

        const fileName = path.basename(filePath);
        const ext = path.extname(fileName);

        const fileBuffer = await readFile(filePath);
        if (fileBuffer) {

            const content = fileBuffer.toString();

            context.componentFiles.push({
                path: filePath,
                name: fileName.substr(0, fileName.length - ext.length),
                content: content,
                dom: parseDOM(content, { xmlMode: true, withStartIndices: true, withEndIndices: true }),
                links: [],
                globalStyles: [],
                templates: {},
                errors: []
            });
        }

    }
}

function transformFile(context: HandlerContext, componentFile: ComponentFileInfo) {

    for (const node of componentFile.dom) {

        executeComponentNode("beforeTransform", context, componentFile, node);
        executeComponentNode("transform", context, componentFile, node);

        if (node.type) {
            const nodeType = node.type.toLowerCase();

            switch (nodeType) {
                case "style":
                    executeComponentNode("beforeTransformStyle", context, componentFile, node as Element);
                    executeComponentNode("transformStyle", context, componentFile, node as Element);
                    executeComponentNode("afterTransformStyle", context, componentFile, node as Element);
                    break;
                case "link":
                    executeComponentNode("beforeTransformLink", context, componentFile, node as Element);
                    executeComponentNode("transformLink", context, componentFile, node as Element);
                    executeComponentNode("afterTransformLink", context, componentFile, node as Element);
                    break;
                case "script":
                    executeComponentNode("beforeTransformScript", context, componentFile, node as Element);
                    executeComponentNode("transformScript", context, componentFile, node as Element);
                    executeComponentNode("afterTransformScript", context, componentFile, node as Element);
                    break;
                case "template":
                    executeComponentNode("beforeTransformTemplate", context, componentFile, node as Element);
                    executeComponentNode("transformTemplate", context, componentFile, node as Element);
                    executeComponentNode("afterTransformTemplate", context, componentFile, node as Element);
                    break;
            }

        }

        executeComponentNode("afterTransform", context, componentFile, node);
    }
}

function parseFile(context: HandlerContext, componentFile: ComponentFileInfo) {

    for (const node of componentFile.dom) {

        if (node.type) {
            const nodeType = node.type.toLowerCase();

            switch (nodeType) {
                case "link":
                case "style":
                    executeComponentNode("beforeParseStyle", context, componentFile, node as Element);
                    executeComponentNode("parseStyle", context, componentFile, node as Element);
                    executeComponentNode("afterParseStyle", context, componentFile, node as Element);
                    break;
                case "script":
                    executeComponentNode("beforeParseScript", context, componentFile, node as Element);
                    executeComponentNode("parseScript", context, componentFile, node as Element);
                    executeComponentNode("afterParseScript", context, componentFile, node as Element);
                    break;
                case "template":
                    executeComponentNode("beforeParseTemplate", context, componentFile, node as Element);
                    executeComponentNode("parseTemplate", context, componentFile, node as Element);
                    executeComponentNode("afterParseTemplate", context, componentFile, node as Element);
                    break;
            }
        }
    }
}

function codegenFile(context: HandlerContext, componentFile: ComponentFileInfo) {
    executeComponent("beforeCodegen", context, componentFile);
    executeComponent("codegen", context, componentFile);
    executeComponent("afterCodegen", context, componentFile);
}



function execute(handlerName: PluginHandlers, context: HandlerContext) {
    for (const plugin of context.plugins) {
        const handler = plugin[handlerName];
        if (handler && handler(context)) {
            return true;
        }
    }
    return false;
}
function executeComponent(handlerName: PluginComponentHandlers, context: HandlerContext, componentFile: ComponentFileInfo) {

    for (const plugin of context.plugins) {
        const handler = plugin[handlerName];
        if (handler && handler(context, componentFile)) {
            return true;
        }
    }
    return false;
}
function executeComponentNode<T extends keyof PluginHandlerMap>(handlerName: T, context: HandlerContext, componentFile: ComponentFileInfo, element: PluginHandlerMap[T]) {

    for (const plugin of context.plugins) {
        const handler = plugin[handlerName];
        if (handler && handler(context, componentFile, element as any)) {
            return true;
        }
    }


    return false;
}

export function extractExpressions(content: string) {

    const expressions: string[] = [];

    const regex = /([^\\]|^)\${([^}]*)}/gm;

    let m;

    while ((m = regex.exec(content)) !== null) {
        // This is necessary to avoid infinite loops with zero-width matches
        if (m.index === regex.lastIndex) {
            regex.lastIndex++;
        }

        expressions.push(m[2]);
    }

    return expressions;
}

export function parseExpressions(content: string) {

    const results = parseSync("`" + content + "`", {
        babelrc: false,
        configFile: false
    });
    if (!results) return;

    const state: ExpressionInfo = {
        content,
        hasTokens: false,
        externalReferences: []
    };

    traverse(results, {
        TemplateLiteral(path) {
            if (path.node.expressions.length > 0) {
                state.hasTokens = true;
            }
        },
        Identifier(path, state: ExpressionInfo) {
            const name = path.node.name;
            if (!path.scope.hasBinding(name)) {
                state.externalReferences.push(name);
            }
        },
    }, undefined, state);

    return state;
}


const defaultHandlers: Handler[] = [

];


