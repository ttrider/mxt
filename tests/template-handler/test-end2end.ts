import path from "path";
import { ComponentFile } from "../../src/component-file";
import processComponentFile from "../../src/defaultHandlers/file-handler";
import processTemplate from "../../src/defaultHandlers/template-handler";
import { formatComponentFileObject } from "../utils";


describe("msx-end2end", () => {
    test("t00", async () => {
        const { componentFile } = await setup("t00");
        const cf = formatComponentFileObject(componentFile);
        expect(cf).toMatchSnapshot();
    });

});




async function setup(fileName: string, id: string = "t01") {

    const inputFile = path.resolve(__dirname, "..", "resources", fileName + ".html");
    const componentFile = await ComponentFile.fromFile(inputFile);

    processComponentFile(componentFile);

    for (const tid in componentFile.templates) {
        if (componentFile.templates.hasOwnProperty(tid)) {
            const template = componentFile.templates[tid];
            await processTemplate(componentFile, template);
        }
    }

    const component = componentFile.components[id];

    const formatted = formatComponentFileObject(componentFile);

    return {
        componentFile,
        component,
        formatted
    }
}