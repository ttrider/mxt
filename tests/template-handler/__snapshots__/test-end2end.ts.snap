// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`msx-end2end t0 1`] = `
"{
    \\"problems\\": [],
    \\"templates\\": {
        \\"t01\\": {
            \\"type\\": \\"tag\\",
            \\"startIndex\\": 0,
            \\"endIndex\\": 67,
            \\"children\\": [
                {
                    \\"type\\": \\"text\\",
                    \\"startIndex\\": 30,
                    \\"endIndex\\": 34,
                    \\"data\\": \\"\\\\n    \\"
                },
                {
                    \\"type\\": \\"tag\\",
                    \\"startIndex\\": 35,
                    \\"endIndex\\": 55,
                    \\"children\\": [
                        {
                            \\"type\\": \\"text\\",
                            \\"startIndex\\": 40,
                            \\"endIndex\\": 49,
                            \\"data\\": \\"Hello MXT!\\"
                        }
                    ],
                    \\"name\\": \\"div\\",
                    \\"attribs\\": {}
                },
                {
                    \\"type\\": \\"text\\",
                    \\"startIndex\\": 56,
                    \\"endIndex\\": 56,
                    \\"data\\": \\"\\\\n\\"
                }
            ],
            \\"name\\": \\"template\\",
            \\"attribs\\": {
                \\"id\\": \\"t01\\",
                \\"type\\": \\"mxt\\"
            }
        }
    },
    \\"scripts\\": [],
    \\"components\\": {
        \\"t01\\": {
            \\"parts\\": {},
            \\"importParts\\": {},
            \\"partset\\": {
                \\"p0\\": {
                    \\"id\\": \\"p0\\",
                    \\"elements\\": [
                        {
                            \\"type\\": \\"text\\",
                            \\"startIndex\\": 30,
                            \\"endIndex\\": 34,
                            \\"data\\": \\"\\\\n    \\"
                        },
                        {
                            \\"type\\": \\"tag\\",
                            \\"startIndex\\": 35,
                            \\"endIndex\\": 55,
                            \\"children\\": [
                                {
                                    \\"type\\": \\"text\\",
                                    \\"startIndex\\": 40,
                                    \\"endIndex\\": 49,
                                    \\"data\\": \\"Hello MXT!\\"
                                }
                            ],
                            \\"name\\": \\"div\\",
                            \\"attribs\\": {}
                        },
                        {
                            \\"type\\": \\"text\\",
                            \\"startIndex\\": 56,
                            \\"endIndex\\": 56,
                            \\"data\\": \\"\\\\n\\"
                        }
                    ],
                    \\"attachTo\\": []
                }
            },
            \\"id\\": \\"t01\\",
            \\"rootPart\\": {
                \\"partId\\": \\"p0\\"
            }
        }
    },
    \\"initStatements\\": [],
    \\"componentStatements\\": [],
    \\"imports\\": [],
    \\"content\\": \\"<template id=\\\\\\"t01\\\\\\" type=\\\\\\"mxt\\\\\\">\\\\n    <div>Hello MXT!</div>\\\\n</template>\\",
    \\"dom\\": [
        {
            \\"type\\": \\"tag\\",
            \\"startIndex\\": 0,
            \\"endIndex\\": 67,
            \\"children\\": [
                {
                    \\"type\\": \\"text\\",
                    \\"startIndex\\": 30,
                    \\"endIndex\\": 34,
                    \\"data\\": \\"\\\\n    \\"
                },
                {
                    \\"type\\": \\"tag\\",
                    \\"startIndex\\": 35,
                    \\"endIndex\\": 55,
                    \\"children\\": [
                        {
                            \\"type\\": \\"text\\",
                            \\"startIndex\\": 40,
                            \\"endIndex\\": 49,
                            \\"data\\": \\"Hello MXT!\\"
                        }
                    ],
                    \\"name\\": \\"div\\",
                    \\"attribs\\": {}
                },
                {
                    \\"type\\": \\"text\\",
                    \\"startIndex\\": 56,
                    \\"endIndex\\": 56,
                    \\"data\\": \\"\\\\n\\"
                }
            ],
            \\"name\\": \\"template\\",
            \\"attribs\\": {
                \\"id\\": \\"t01\\",
                \\"type\\": \\"mxt\\"
            }
        }
    ],
    \\"filePath\\": \\"/Users/vladyang/src/github/mxt/tests/resources/t0.html\\"
}"
`;

exports[`msx-end2end t1 1`] = `
"{
    \\"problems\\": [],
    \\"templates\\": {
        \\"t01\\": {
            \\"type\\": \\"tag\\",
            \\"startIndex\\": 0,
            \\"endIndex\\": 94,
            \\"children\\": [
                {
                    \\"type\\": \\"text\\",
                    \\"startIndex\\": 30,
                    \\"endIndex\\": 34,
                    \\"data\\": \\"\\\\n    \\"
                },
                {
                    \\"type\\": \\"tag\\",
                    \\"startIndex\\": 35,
                    \\"endIndex\\": 82,
                    \\"children\\": [
                        {
                            \\"type\\": \\"text\\",
                            \\"startIndex\\": 67,
                            \\"endIndex\\": 76,
                            \\"data\\": \\"Hello MXT!\\"
                        }
                    ],
                    \\"name\\": \\"div\\",
                    \\"attribs\\": {
                        \\"id\\": \\"tagid_1\\"
                    }
                },
                {
                    \\"type\\": \\"text\\",
                    \\"startIndex\\": 83,
                    \\"endIndex\\": 83,
                    \\"data\\": \\"\\\\n\\"
                }
            ],
            \\"name\\": \\"template\\",
            \\"attribs\\": {
                \\"id\\": \\"t01\\",
                \\"type\\": \\"mxt\\"
            }
        }
    },
    \\"scripts\\": [],
    \\"components\\": {
        \\"t01\\": {
            \\"parts\\": {},
            \\"importParts\\": {},
            \\"partset\\": {
                \\"p1\\": {
                    \\"id\\": \\"p1\\",
                    \\"elements\\": [
                        {
                            \\"type\\": \\"text\\",
                            \\"startIndex\\": 30,
                            \\"endIndex\\": 34,
                            \\"data\\": \\"\\\\n    \\"
                        },
                        {
                            \\"type\\": \\"tag\\",
                            \\"startIndex\\": 35,
                            \\"endIndex\\": 82,
                            \\"children\\": [
                                {
                                    \\"type\\": \\"text\\",
                                    \\"startIndex\\": 67,
                                    \\"endIndex\\": 76,
                                    \\"data\\": \\"Hello MXT!\\"
                                }
                            ],
                            \\"name\\": \\"div\\",
                            \\"attribs\\": {
                                \\"id\\": \\"tagid_1\\"
                            }
                        },
                        {
                            \\"type\\": \\"text\\",
                            \\"startIndex\\": 83,
                            \\"endIndex\\": 83,
                            \\"data\\": \\"\\\\n\\"
                        }
                    ],
                    \\"attachTo\\": [
                        {
                            \\"attributes\\": {
                                \\"style\\": {
                                    \\"content\\": \\"color: \${newColor}\\",
                                    \\"externalReferences\\": [
                                        \\"newColor\\"
                                    ],
                                    \\"tokens\\": [
                                        {
                                            \\"start\\": 7,
                                            \\"end\\": 18,
                                            \\"content\\": \\"newColor\\",
                                            \\"token\\": \\"\${newColor}\\"
                                        }
                                    ],
                                    \\"attributeName\\": \\"style\\"
                                }
                            },
                            \\"events\\": {},
                            \\"id\\": \\"tagid_1\\",
                            \\"originalId\\": \\"\\",
                            \\"embeddedParts\\": []
                        }
                    ]
                }
            },
            \\"id\\": \\"t01\\",
            \\"rootPart\\": {
                \\"partId\\": \\"p1\\"
            }
        }
    },
    \\"initStatements\\": [],
    \\"componentStatements\\": [],
    \\"imports\\": [],
    \\"content\\": \\"<template id=\\\\\\"t01\\\\\\" type=\\\\\\"mxt\\\\\\">\\\\n    <div style=\\\\\\"color: \${newColor}\\\\\\">Hello MXT!</div>\\\\n</template>\\",
    \\"dom\\": [
        {
            \\"type\\": \\"tag\\",
            \\"startIndex\\": 0,
            \\"endIndex\\": 94,
            \\"children\\": [
                {
                    \\"type\\": \\"text\\",
                    \\"startIndex\\": 30,
                    \\"endIndex\\": 34,
                    \\"data\\": \\"\\\\n    \\"
                },
                {
                    \\"type\\": \\"tag\\",
                    \\"startIndex\\": 35,
                    \\"endIndex\\": 82,
                    \\"children\\": [
                        {
                            \\"type\\": \\"text\\",
                            \\"startIndex\\": 67,
                            \\"endIndex\\": 76,
                            \\"data\\": \\"Hello MXT!\\"
                        }
                    ],
                    \\"name\\": \\"div\\",
                    \\"attribs\\": {
                        \\"id\\": \\"tagid_1\\"
                    }
                },
                {
                    \\"type\\": \\"text\\",
                    \\"startIndex\\": 83,
                    \\"endIndex\\": 83,
                    \\"data\\": \\"\\\\n\\"
                }
            ],
            \\"name\\": \\"template\\",
            \\"attribs\\": {
                \\"id\\": \\"t01\\",
                \\"type\\": \\"mxt\\"
            }
        }
    ],
    \\"filePath\\": \\"/Users/vladyang/src/github/mxt/tests/resources/t1.html\\"
}"
`;

exports[`msx-end2end t2 1`] = `
"{
    \\"problems\\": [],
    \\"templates\\": {
        \\"t01\\": {
            \\"type\\": \\"tag\\",
            \\"startIndex\\": 0,
            \\"endIndex\\": 177,
            \\"children\\": [
                {
                    \\"type\\": \\"text\\",
                    \\"startIndex\\": 30,
                    \\"endIndex\\": 34,
                    \\"data\\": \\"\\\\n    \\"
                },
                {
                    \\"type\\": \\"tag\\",
                    \\"startIndex\\": 35,
                    \\"endIndex\\": 165,
                    \\"children\\": [
                        {
                            \\"type\\": \\"text\\",
                            \\"startIndex\\": 141,
                            \\"endIndex\\": 159,
                            \\"data\\": \\"\\\\n        Hello MXT!\\"
                        }
                    ],
                    \\"name\\": \\"div\\",
                    \\"attribs\\": {
                        \\"id\\": \\"tagid_2\\"
                    }
                },
                {
                    \\"type\\": \\"text\\",
                    \\"startIndex\\": 166,
                    \\"endIndex\\": 166,
                    \\"data\\": \\"\\\\n\\"
                }
            ],
            \\"name\\": \\"template\\",
            \\"attribs\\": {
                \\"id\\": \\"t01\\",
                \\"type\\": \\"mxt\\"
            }
        }
    },
    \\"scripts\\": [],
    \\"components\\": {
        \\"t01\\": {
            \\"parts\\": {},
            \\"importParts\\": {},
            \\"partset\\": {
                \\"p2\\": {
                    \\"id\\": \\"p2\\",
                    \\"elements\\": [
                        {
                            \\"type\\": \\"text\\",
                            \\"startIndex\\": 30,
                            \\"endIndex\\": 34,
                            \\"data\\": \\"\\\\n    \\"
                        },
                        {
                            \\"type\\": \\"tag\\",
                            \\"startIndex\\": 35,
                            \\"endIndex\\": 165,
                            \\"children\\": [
                                {
                                    \\"type\\": \\"text\\",
                                    \\"startIndex\\": 141,
                                    \\"endIndex\\": 159,
                                    \\"data\\": \\"\\\\n        Hello MXT!\\"
                                }
                            ],
                            \\"name\\": \\"div\\",
                            \\"attribs\\": {
                                \\"id\\": \\"tagid_2\\"
                            }
                        },
                        {
                            \\"type\\": \\"text\\",
                            \\"startIndex\\": 166,
                            \\"endIndex\\": 166,
                            \\"data\\": \\"\\\\n\\"
                        }
                    ],
                    \\"attachTo\\": [
                        {
                            \\"attributes\\": {
                                \\"style\\": {
                                    \\"content\\": \\"color: \${newColor}\\",
                                    \\"externalReferences\\": [
                                        \\"newColor\\"
                                    ],
                                    \\"tokens\\": [
                                        {
                                            \\"start\\": 7,
                                            \\"end\\": 18,
                                            \\"content\\": \\"newColor\\",
                                            \\"token\\": \\"\${newColor}\\"
                                        }
                                    ],
                                    \\"attributeName\\": \\"style\\"
                                }
                            },
                            \\"events\\": {
                                \\"click\\": {
                                    \\"name\\": \\"click\\",
                                    \\"handler\\": \\"doClick\\",
                                    \\"preventDefault\\": true
                                },
                                \\"zoom\\": {
                                    \\"name\\": \\"zoom\\",
                                    \\"handler\\": \\"something.zoom\\",
                                    \\"externalReferences\\": [
                                        \\"something\\"
                                    ]
                                }
                            },
                            \\"id\\": \\"tagid_2\\",
                            \\"originalId\\": \\"\\",
                            \\"embeddedParts\\": []
                        }
                    ]
                }
            },
            \\"id\\": \\"t01\\",
            \\"rootPart\\": {
                \\"partId\\": \\"p2\\"
            }
        }
    },
    \\"initStatements\\": [],
    \\"componentStatements\\": [],
    \\"imports\\": [],
    \\"content\\": \\"<template id=\\\\\\"t01\\\\\\" type=\\\\\\"mxt\\\\\\">\\\\n    <div style=\\\\\\"color: \${newColor}\\\\\\" mxt.click=\\\\\\"doClick\\\\\\" mxt.zoom=\\\\\\"\${something.zoom}\\\\\\" mxt.click.preventDefault>\\\\n        Hello MXT!</div>\\\\n</template>\\",
    \\"dom\\": [
        {
            \\"type\\": \\"tag\\",
            \\"startIndex\\": 0,
            \\"endIndex\\": 177,
            \\"children\\": [
                {
                    \\"type\\": \\"text\\",
                    \\"startIndex\\": 30,
                    \\"endIndex\\": 34,
                    \\"data\\": \\"\\\\n    \\"
                },
                {
                    \\"type\\": \\"tag\\",
                    \\"startIndex\\": 35,
                    \\"endIndex\\": 165,
                    \\"children\\": [
                        {
                            \\"type\\": \\"text\\",
                            \\"startIndex\\": 141,
                            \\"endIndex\\": 159,
                            \\"data\\": \\"\\\\n        Hello MXT!\\"
                        }
                    ],
                    \\"name\\": \\"div\\",
                    \\"attribs\\": {
                        \\"id\\": \\"tagid_2\\"
                    }
                },
                {
                    \\"type\\": \\"text\\",
                    \\"startIndex\\": 166,
                    \\"endIndex\\": 166,
                    \\"data\\": \\"\\\\n\\"
                }
            ],
            \\"name\\": \\"template\\",
            \\"attribs\\": {
                \\"id\\": \\"t01\\",
                \\"type\\": \\"mxt\\"
            }
        }
    ],
    \\"filePath\\": \\"/Users/vladyang/src/github/mxt/tests/resources/t2.html\\"
}"
`;
